src/scheduler/gc_work.rs

```diff
@@ -323,6 +357,16 @@ impl<E: ProcessEdgesWork> ObjectTracerContext<E::VM> for ProcessEdgesWorkTracerC
     fn with_tracer<R, F>(&self, worker: &mut GCWorker<E::VM>, func: F) -> R
     where
         F: FnOnce(&mut Self::TracerType) -> R,
+    {
+        self.with_tracer_and_worker(worker, |tracer, _| func(tracer))
+    }
+}
+
+impl<E: ProcessEdgesWork> ProcessEdgesWorkTracerContext<E> {
+    // This Also exposes worker to the callback function. This is not a public method.
+    fn with_tracer_and_worker<R, F>(&self, worker: &mut GCWorker<E::VM>, func: F) -> R
+    where
+        F: FnOnce(&mut ProcessEdgesWorkTracer<E>, &mut GCWorker<E::VM>) -> R,
```

Comment:
Strictly speaking, it is unsafe to give the user access to both the tracer and the worker.  The tracer is a closure that calls `trace_object` underneath, and the `worker` is also one of the arguments of `trace_object`.  I planned to forbid this by adding a lifetime annotation, but I couldn't do it without the General Associated Type (GAT) feature before Rust 1.65.

See: https://github.com/mmtk/mmtk-core/blob/813fa99fbf5b5af603d58bd3b54576c141188c37/src/vm/scanning.rs#L53

Reply:
Another problem with this is that VM bindings can use `ObjectTracerContext` to get temporary access to `trace_object` during `Scanning::process_weak_refs` and `Scanning::forward_weak_refs`.  Adding a `with_tracer_and_worker` will allow `ScanObjects` to record object sizes, but not for VM-specific weak reference processing.

Reply:
`with_tracer` takes a mutable reference of `worker`, so I cannot use `worker` any more in `ScanObjects::do_work_common`. We need to a way to get the `worker` back from the tracer. Another option is to expose some methods from `ObjectTracer` so we can get the worker or directly increase the counter.

Reply:
> Another problem with this is that VM bindings can use `ObjectTracerContext` to get temporary access to `trace_object` during `Scanning::process_weak_refs` and `Scanning::forward_weak_refs`. Adding a `with_tracer_and_worker` will allow `ScanObjects` to record object sizes, but not for VM-specific weak reference processing.

The remaining question is how to count live bytes for weak references. As the weak reference processing is totally at the binding side, we may have to expose a method and let the binding to call it when they scan weak refs.

Reply:
> The remaining question is how to count live bytes for weak references. As the weak reference processing is totally at the binding side, we may have to expose a method and let the binding to call it when they scan weak refs.

If we count the object size at `ScanObjectsWork`, it will not be a problem.  The binding keeps objects alive during the weak reference processing stage using `trace_object` (via `ObjectTracer::trace_object`, which is implemented with `ProcessEdgesWork::trace_object` underneath and eventually gets to the space-specific `trace_object`).  Any object visited by `trace_object` will eventually be visited by `ScanObjectsWork`.